app/src/main/java/com/epfl/drawyourpath/userProfile/cache/UserModelCached.kt

```diff
@@ -111,17 +116,44 @@ class UserModelCached(application: Application) : AndroidViewModel(application)
         checkCurrentUser(false)
         setUserId(userId)
         return database.getUserData(userId).thenApplyAsync { userData ->
-            val runs = RunEntity.fromRunsToEntities(userData.userId ?: userId, userData.runs ?: listOf())
+            val runs =
+                RunEntity.fromRunsToEntities(userData.userId ?: userId, userData.runs ?: listOf())
             userCache.insertAll(
                 UserEntity(userData, userId),
                 userData.dailyGoals?.map { DailyGoalEntity(it, userId) } ?: listOf(),
-                listOf(), // TODO replace with userData.milestones when there is one
+                fromMilestoneDataToEntity(userId = userId, milestonesData = userData.milestones),
                 runs.map { it.first },
                 runs.map { it.second }.flatten(),
             )
         }
     }
 
+    /**
+     * Helper function to transform MilestoneData into Milestone entities
+     * @param milestoneData list of milestone data
+     * @param userId of the user who passed the milestones
+     * @return a list of milestone entities
+     */
+    private fun fromMilestoneDataToEntity(
+        userId: String,
+        milestonesData: List<MilestoneData>?,
+    ): List<MilestoneEntity> {
+        if (milestonesData.isNullOrEmpty()) {
+            return emptyList()
+        }
+        val list = mutableListOf<MilestoneEntity>()
+        milestonesData.forEach { milestone ->
+            list.add(
+                MilestoneEntity(
+                    userId = userId,
+                    milestone = Utils.getStringFromALL_CAPS(milestone.milestone!!.name),
+                    date = milestone.date!!.toEpochDay(),
+                ),
+            )
+        }
+        return list

```

Comment:
This can be done in one line:

```kotlin
return milestonesData?.map { MilestoneEntity(userId, Milestone(it.milestone!!, it.date!!)) } ?: listOf()
```